File: java/DivideTwoIntegers.java
//when programming in java, you have to notice abs(MIN_VALUE) might cause overflow
//I don't want to use Long to solve it.
public class Solution {
    public int divide(int dividend, int divisor) {
        int negative = 1;
        int result = 0;
        if (divisor == 0) {
            return Integer.MAX_VALUE;
        }
        
        if (dividend == Integer.MIN_VALUE) {
            if (divisor == -1) {
                return Integer.MAX_VALUE;   
            }
            if (divisor == Integer.MIN_VALUE) {
                return 1;
            }
            dividend += Math.abs(divisor);
            result += 1;
        }
        
        if (divisor == Integer.MIN_VALUE) {
            return 0;
        }
        if (dividend < 0 && divisor > 0 || dividend > 0 && divisor < 0) {
            negative = -1;
        }
        dividend = Math.abs(dividend);
        divisor = Math.abs(divisor);
        
        int tempDivisor = divisor;
        int i = 1;
        while (dividend >= divisor) {
            while (dividend - tempDivisor >= tempDivisor) {
                tempDivisor = tempDivisor + tempDivisor;
                i = i + i;
            }
            dividend = dividend - tempDivisor;
            tempDivisor = divisor;
            result += i;
            i = 1;
        }
        return negative * result;
    }
}

File: archive/python/math/DivideTwoIntegers.py
class Solution:
    # @return an integer
    def divide(self, dividend, divisor):
        INT_MAX = 2147483647
        INT_MIN = -2147483648
        
        if dividend == INT_MIN and divisor == -1:
            return INT_MAX
            
        if dividend == INT_MAX and divisor == 1:
            return INT_MAX
            
        if divisor == 0:
            return INT_MAX
        
        negative = 1
        if (dividend > 0 and divisor < 0) or (dividend < 0 and divisor > 0):
            negative = -1
        
        dividend = abs(dividend)
        divisor = abs(divisor)
        
        d = dividend
        s = divisor
        result = 0
        i = 1
        while d >= divisor:
            while s + s < d:
                s = s + s
                i = i + i
            result += i
            i = 1
            d = d - s
            s = divisor
        return negative * result

if __name__ == "__main__":
    solution = Solution()
    print solution.divide(31, 3)
